x/mongo/driver/auth/oidc.go

```diff
@@ -130,38 +130,117 @@ func newOIDCAuthenticator(cred *Cred, httpClient *http.Client) (Authenticator, e
 	return oa, nil
 }
 
+func createAllowedHostsPatterns(hosts []string) []string {
+	for i := range hosts {
+		hosts[i] = strings.ReplaceAll(hosts[i], ".", "[.]")
+		hosts[i] = strings.ReplaceAll(hosts[i], "*", ".*")
+		hosts[i] = "^" + hosts[i] + "(:\\d+)?$"
+	}
+	return hosts
+}
+
+func (oa *OIDCAuthenticator) getAllowedHosts() []string {
+	// we cache allowed hosts so that we do not need to convert globs to patterns
+	// every time we authenticate.
+	if oa.allowedHosts != nil {
+		return *oa.allowedHosts
+	}
+	var ret []string
+	allowedHosts, ok := oa.AuthMechanismProperties[allowedHostsProp]
+	if ok {
+		ret = strings.Split(allowedHosts, ",")
+	} else {
+		ret = make([]string, len(defaultAllowedHosts))
+		copy(ret, defaultAllowedHosts)
+	}
+	ret = createAllowedHostsPatterns(ret)
+	oa.allowedHosts = &ret
```

Comment:
Optional: Consider initializing `allowedHosts` in `newOIDCAuthenticator`, which would allow compiling the `ALLOWED_HOSTS` patterns as a slice of `*regexp.Regexp` there. That would also allow pre-compiling the default allowed hosts patterns, avoiding the CodeQL warnings and catching any regex compilation errors at initialization time.

For example:
```go
func newOIDCAuthenticator(...) (Authenticator, error) {
	// ...
	var allowedHosts []*regexp.Regexp
	if v, ok := cred.Props[allowedHostsProp]; ok {
		patterns := strings.Split(v, ",")
		for _, pattern := range patterns {
			r, err := regexp.Compile(pattern)
			if err != nil {
				return nil, err
			}
			allowedHosts = append(allowedHosts, r)
		}
	} else {
		allowedHosts = defaultAllowedHosts
	}
	oa := &OIDCAuthenticator{
		userName:                cred.Username,
		httpClient:              httpClient,
		AuthMechanismProperties: cred.Props,
		OIDCMachineCallback:     cred.OIDCMachineCallback,
		OIDCHumanCallback:       cred.OIDCHumanCallback,
		allowedHosts:            allowedHosts,
	}
	return oa, nil
}
```
Then `defaultAllowedHosts` becomes:
```go
var defaultAllowedHosts = []*regexp.Regexp{
	regexp.MustCompile("^.*[.]mongodb[.]net(:\\d+)?$"),
	regexp.MustCompile("^.*[.]mongodb-qa[.]net(:\\d+)?$"),
	regexp.MustCompile("^.*[.]mongodb-dev[.]net(:\\d+)?$"),
	regexp.MustCompile("^.*[.]mongodbgov[.]net(:\\d+)?$"),
	regexp.MustCompile("^localhost(:\\d+)?$"),
	regexp.MustCompile("^127[.]0[.]0[.]1(:\\d+)?$"),
	regexp.MustCompile("^::1(:\\d+)?$"),
}
```

Reply:
Ok, I've adapted most of this, but kept the logic abstracted to a separate method.